### ISSUE TITLE
Dependency Conflict: duplicate classes"org.apache.http.message.BasicHeaderValueParser" in different JARs, have different implementations 

### ISSUE THREAD

**Issue by HelloCoCooo** [ontop/ontop#287]
Hi, in **ontop-ontop-3.0.0-beta-2** (**\engine\system\sql\owlapi** module), duplicate classes with the same fully-qualified name _**org.apache.http.message.BasicHeaderValueParser**_ are included in two different libraries, i.e., **org.apache.httpcomponents:httpcore-osgi:4.4.4** and **org.apache.httpcomponents:httpcore:4.2.4**.

According to _"first declaration wins"_ class loading strategy, only this class in **_org.apache.httpcomponents:httpcore-osgi:4.4.4_** can be loaded, and that in _**org.apache.httpcomponents:httpcore:4.2.4**_ will be shadowed.

By further analyzing, your project expects to invoke method **_<org.apache.http.message.BasicHeaderValueParser: org.apache.http.NameValuePair parseNameValuePair(org.apache.http.util.CharArrayBuffer,org.apache.http.message.ParserCursor)>_**  in _**org.apache.httpcomponents:httpcore:4.2.4**_. As it has been shadowed, so that this method defined in **_org.apache.httpcomponents:httpcore-osgi:4.4.4_** is actually forced to be referenced via the following invocation path:
```
<it.unibz.inf.ontop.injection.impl.OntopSQLOWLAPIConfigurationImpl$OntopSQLOWLAPIBuilderMixin: it.unibz.inf.ontop.injection.OntopSQLOWLAPIConfiguration$Builder ontologyFile(java.net.URL)> D:\testcase\TestProject\ontop-ontop-3.0.0-beta-2\engine\system\sql\owlapi\target\classes
<it.unibz.inf.ontop.injection.impl.OntopMappingOntologyBuilders$StandardMappingOntologyBuilderFragment: it.unibz.inf.ontop.injection.OntopMappingOntologyConfiguration$Builder ontologyFile(java.net.URL)> D:\cEnvironment\repository\it\unibz\inf\ontop\ontop-mapping-owlapi\3.0.0-beta-2\ontop-mapping-owlapi-3.0.0-beta-2.jar
<org.apache.thrift.server.TThreadedSelectorServer$SelectorThread: void run()> D:\cEnvironment\repository\org\apache\thrift\libthrift\0.9.3\libthrift-0.9.3.jar
<org.apache.http.message.BasicHeaderElementIterator: boolean hasNext()> 
<org.apache.http.message.BasicHeaderElementIterator: void parseNextElement()> 
<org.apache.http.message.BasicHeaderValueParser: org.apache.http.HeaderElement parseHeaderElement(org.apache.http.util.CharArrayBuffer,org.apache.http.message.ParserCursor)> 
<org.apache.http.message.BasicHeaderValueParser: org.apache.http.NameValuePair[] parseParameters(org.apache.http.util.CharArrayBuffer,org.apache.http.message.ParserCursor)> 

```
Although both of these two conflicting classes contain the referenced methods (with the same signature), they have different implementations. This issue will not lead to runtime crashes, but it can introduce inconsistent semantic hehavior by changing the control flows and data flows.

**_Workaround solution:_**
An easy way to workaround the problem is **_reversing the declaration order_** of these two libraries in pom file. 
Then, according to **_"first declaration wins"_** class loading strategy, class _**org.apache.http.message.BasicHeaderValueParser**_ in _**org.apache.httpcomponents:httpcore:4.2.4**_ can be loaded (the version that **ontop** expects to reference by static analysis).
This fix will not affect other libraries or class, except the above duplicate class.

Dependency tree--
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ ontop-system-sql-owlapi ---
[INFO] it.unibz.inf.ontop:ontop-system-sql-owlapi:jar:3.0.0-beta-2
**.....**
[INFO] |  +- it.unibz.inf.ontop:ontop-reformulation-sql:jar:3.0.0-beta-2:compile
[INFO] |  |  \- it.unibz.inf.ontop:ontop-rdb:jar:3.0.0-beta-2:compile
[INFO] |  +- it.unibz.inf.ontop:ontop-mapping-sql-all:jar:3.0.0-beta-2:compile
[INFO] |  |  +- it.unibz.inf.ontop:ontop-mapping-sql-core:jar:3.0.0-beta-2:compile
[INFO] |  |  +- it.unibz.inf.ontop:ontop-mapping-native:jar:3.0.0-beta-2:compile
[INFO] |  |  +- it.unibz.inf.ontop:ontop-mapping-r2rml:jar:3.0.0-beta-2:compile
[INFO] |  |  |  +- org.eclipse.rdf4j:rdf4j-rio-turtle:jar:2.2.2:compile
[INFO] |  |  |  |  +- org.eclipse.rdf4j:rdf4j-rio-datatypes:jar:2.2.2:runtime
[INFO] |  |  |  |  \- org.eclipse.rdf4j:rdf4j-rio-languages:jar:2.2.2:runtime
[INFO] |  |  |  +- eu.optique-project:r2rml-api-core:jar:0.6.0:compile
[INFO] |  |  |  +- eu.optique-project:r2rml-api-rdf4j-binding:jar:0.6.0:compile
[INFO] |  |  |  |  \- org.apache.commons:commons-rdf-rdf4j:jar:0.3.0-incubating:compile
[INFO] |  |  |  |     +- org.apache.commons:commons-rdf-simple:jar:0.3.0-incubating:compile
[INFO] |  |  |  |     +- org.eclipse.rdf4j:rdf4j-repository-api:jar:2.2.2:compile
[INFO] |  |  |  |     +- org.eclipse.rdf4j:rdf4j-rio-ntriples:jar:2.2.2:compile
[INFO] |  |  |  |     +- org.eclipse.rdf4j:rdf4j-rio-nquads:jar:2.1.1:compile
[INFO] |  |  |  |     +- org.eclipse.rdf4j:rdf4j-rio-rdfxml:jar:2.2.2:compile
[INFO] |  |  |  |     +- org.eclipse.rdf4j:rdf4j-rio-trig:jar:2.1.1:compile
[INFO] |  |  |  |     +- org.eclipse.rdf4j:rdf4j-rio-jsonld:jar:2.1.1:compile
[INFO] |  |  |  |     +- org.eclipse.rdf4j:rdf4j-sail-memory:jar:2.1.1:compile
[INFO] |  |  |  |     |  +- org.eclipse.rdf4j:rdf4j-sail-base:jar:2.1.1:compile
[INFO] |  |  |  |     |  +- org.eclipse.rdf4j:rdf4j-sail-inferencer:jar:2.1.1:compile
[INFO] |  |  |  |     |  |  +- org.eclipse.rdf4j:rdf4j-queryparser-serql:jar:2.1.1:runtime
[INFO] |  |  |  |     |  |  \- org.eclipse.rdf4j:rdf4j-sail-model:jar:2.1.1:compile
[INFO] |  |  |  |     |  \- org.eclipse.rdf4j:rdf4j-queryalgebra-evaluation:jar:2.2.2:compile
[INFO] |  |  |  |     |     +- org.eclipse.rdf4j:rdf4j-repository-sparql:jar:2.2.2:compile
[INFO] |  |  |  |     |     |  \- org.eclipse.rdf4j:rdf4j-queryresultio-sparqlxml:jar:2.2.2:compile
[INFO] |  |  |  |     |     \- org.mapdb:mapdb:jar:1.0.8:compile
[INFO] |  |  |  |     \- org.eclipse.rdf4j:rdf4j-repository-sail:jar:2.1.1:compile
[INFO] |  |  |  |        +- org.eclipse.rdf4j:rdf4j-sail-api:jar:2.2.2:compile
[INFO] |  |  |  |        \- org.eclipse.rdf4j:rdf4j-http-client:jar:2.1.1:compile
[INFO] |  |  |  |           +- org.eclipse.rdf4j:rdf4j-http-protocol:jar:2.1.1:compile
[INFO] |  |  |  |           \- org.eclipse.rdf4j:rdf4j-queryresultio-api:jar:2.2.2:compile
[INFO] |  |  |  \- eu.optique-project:r2rml-api-jena-binding:jar:0.6.0:compile
[INFO] |  |  |     +- org.apache.commons:commons-rdf-jena:jar:0.3.0-incubating:compile
[INFO] |  |  |     |  +- org.apache.jena:jena-osgi:jar:3.1.1:compile
[INFO] |  |  |     |  |  +- org.apache.httpcomponents:httpclient-osgi:jar:4.5.2:compile
[INFO] |  |  |     |  |  +- _**org.apache.httpcomponents:httpcore-osgi:jar:4.4.4:compile**_
[INFO] |  |  |     |  |  +- org.apache.commons:commons-csv:jar:1.3:compile
**....**
[INFO] |  +- net.sourceforge.owlapi:owlapi-api:jar:4.2.8:compile
[INFO] |  |  +- org.tukaani:xz:jar:1.5:compile
[INFO] |  |  +- net.sf.trove4j:trove4j:jar:3.0.3:compile
[INFO] |  |  +- com.google.inject.extensions:guice-multibindings:jar:4.0:compile
[INFO] |  |  \- commons-io:commons-io:jar:2.4:compile
[INFO] |  \- net.sourceforge.owlapi:owlapi-apibinding:jar:4.2.8:compile
[INFO] |     +- net.sourceforge.owlapi:owlapi-impl:jar:4.2.8:compile
[INFO] |     +- net.sourceforge.owlapi:owlapi-parsers:jar:4.2.8:compile
[INFO] |     +- net.sourceforge.owlapi:owlapi-oboformat:jar:4.2.8:compile
[INFO] |     +- net.sourceforge.owlapi:owlapi-tools:jar:4.2.8:compile
[INFO] |     +- net.sourceforge.owlapi:owlapi-fixers:jar:4.2.8:compile
[INFO] |     \- net.sourceforge.owlapi:owlapi-rio:jar:4.2.8:compile
[INFO] |        +- org.openrdf.sesame:sesame-model:jar:2.7.16:compile
[INFO] |        |  \- org.openrdf.sesame:sesame-util:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-api:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-languages:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-datatypes:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-binary:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-n3:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-nquads:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-ntriples:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-rdfjson:jar:2.7.16:compile
[INFO] |        |  \- com.fasterxml.jackson.core:jackson-core:jar:2.2.1:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-rdfxml:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-trix:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-turtle:jar:2.7.16:compile
[INFO] |        +- org.openrdf.sesame:sesame-rio-trig:jar:2.7.16:compile
[INFO] |        +- com.github.jsonld-java:jsonld-java-sesame:jar:0.5.0:compile
[INFO] |        |  \- com.github.jsonld-java:jsonld-java:jar:0.5.0:compile
[INFO] |        |     +- com.fasterxml.jackson.core:jackson-databind:jar:2.3.3:compile
[INFO] |        |     |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.3.0:compile
[INFO] |        |     +- org.apache.httpcomponents:httpclient-cache:jar:4.2.5:compile
[INFO] |        |     +- org.apache.httpcomponents:httpclient:jar:4.2.5:compile
[INFO] |        |     |  +- _**org.apache.httpcomponents:httpcore:jar:4.2.4:compile**_
[INFO] |        |     |  \- commons-codec:commons-codec:jar:1.6:compile
[INFO] |        |     \- org.slf4j:jcl-over-slf4j:jar:1.7.7:compile
[INFO] |        \- org.semarglproject:semargl-sesame:jar:0.6.1:compile
[INFO] |           +- org.semarglproject:semargl-core:jar:0.6.1:compile
[INFO] |           \- org.semarglproject:semargl-rdfa:jar:0.6.1:compile
[INFO] |              \- org.semarglproject:semargl-rdf:jar:0.6.1:compile
**.....**
Thank you very much.
Best,
Coco

**Comment by HelloCoCooo** [comment]
Code snippet of **_:<org.apache.http.message.BasicHeaderValueParser: org.apache.http.NameValuePair parseNameValuePair(org.apache.http.util.CharArrayBuffer,org.apache.http.message.ParserCursor)>_** in _**org.apache.httpcomponents:httpcore-osgi:4.4.4**_ (loaded version):

![a](https://user-images.githubusercontent.com/46306510/55814936-102e5b80-5b22-11e9-97f2-7d7d72774410.png)

Code snippet of **_:<org.apache.http.message.BasicHeaderValueParser: org.apache.http.NameValuePair parseNameValuePair(org.apache.http.util.CharArrayBuffer,org.apache.http.message.ParserCursor)>_** in _**org.apache.httpcomponents:httpcore:4.2.4**_ 
 (shadowed but expected to invoke method):

![b](https://user-images.githubusercontent.com/46306510/55814948-16bcd300-5b22-11e9-82e0-ca329e59a172.png)
![c](https://user-images.githubusercontent.com/46306510/55814952-191f2d00-5b22-11e9-9c01-3c3a847ad47e.png)

As a result, these conflicting method included in **org.apache.httpcomponents:httpcore:4.2.4** deals with different cases, which changes the control flows and data flows. So being forced to use these methods in **org.apache.httpcomponents:httpcore-osgi:4.4.4** may lead to inconsisitent semantic behaviors.

**Comment by HelloCoCooo** [comment]
Thanks for your attention.

**Comment by HelloCoCooo** [comment]
This issue and #288 are fixed by commit [eebcbf8](https://github.com/ontop/ontop/commit/eebcbf8e63fd746faf5f0db75c066a53ad20395d).

So that I close it.
